python_wikibase/data_types/time.py

```diff
@@ -0,0 +1,76 @@
+import collections
+import datetime
+
+from python_wikibase.data_types.data_type import DataType
+
+
+class DataTypeException(BaseException):
+    pass
+
+# Largely inspired from  https://github.com/dahlia/wikidata/blob/master/wikidata/datavalue.py
+
+
+class Time(DataType):
+    def __init__(self, py_wb, api, language):
+        super().__init__(py_wb, api, language)
+        self.value = None
+
+    def __str__(self):
+        return self.value.isoformat()
+
+    def unmarshal(self, data_value):
+        value = data_value["value"]
+
+        # See https://www.mediawiki.org/wiki/Wikibase/DataModel/JSON#time
+        # for property to extract from data_value
+
+        time = value['time']
+        # Strip '+' at beginning of string and 'Z' at the end
+        time = time[1:-1]
+
+        tz = value['timezone']
+        # Don't take in account 'before' and 'after' property
+        precision = value['precision']
+
+        if precision == 11:
+            self.value = datetime.date.fromisoformat(time[:-9])
+        elif precision == 14:
+            val = datetime.datetime.fromisoformat(time)
+            val.replace(tzinfo=datetime.timezone(offset=datetime.timedelta(minutes=tz)))
+            self.value = val
+        return self
+
+    def marshal(self):
+
+        common_properties = {
+            "timezone": 0,
+            "before": 0,
+            "after": 0,
+            "calendarmodel": "http:\/\/www.wikidata.org\/entity\/Q1985727"
```

Comment:
I'm not sure we can hard-code an item from wikidata.org. It will probably not work with a custom instance of wikibase.

Reply:
I found this URL [there](https://github.com/dahlia/wikidata/blob/3bfaee438a571731d71d454f578f3b972b70a770/wikidata/datavalue.py#L163) and [there](https://www.mediawiki.org/wiki/Wikibase/DataModel/JSON#time)

But I fix url, replacing "\/" by "/".

Reply:
For calendarmodel, 2 urls are used:
- https://www.wikidata.org/wiki/Q1985727 (Gregorian calendar)
- https://www.wikidata.org/wiki/Q1985786 (Julian calendar)

I pushed a new commit that extract first url as a more readable constant `GREGORIAN_CALENDAR_WIKIDATA_URL`